game/src/common/mod.rs

```diff
@@ -407,3 +409,103 @@ pub fn checkbox_per_mode(
     }
     Widget::custom_row(filters)
 }
+
+/// This does nothing on native. On web, it modifies the current URL to change the first free
+/// parameter in the HTTP GET params to the specified value, adding it if needed.
+#[allow(unused_variables)]
+pub fn update_url(free_param: &str) -> Result<()> {
+    #[cfg(target_arch = "wasm32")]
+    {
+        let window = web_sys::window().ok_or(anyhow!("no window?"))?;
+        let url = window.location().href().map_err(|err| {
+            anyhow!(err
+                .as_string()
+                .unwrap_or("window.location.href failed".to_string()))
+        })?;
+        let new_url = change_url_free_query_param(url, free_param);
+
+        // Setting window.location.href may seem like the obvious thing to do, but that actually
+        // refreshes the page. This method just changes the URL and doesn't mess up history. See
+        // https://developer.mozilla.org/en-US/docs/Web/API/History_API/Working_with_the_History_API.
+        let history = window.history().map_err(|err| {
```

Comment:
TIL that web-sys Results have a JSValue as their Err case.

Some discussion to change that:
https://github.com/rustwasm/wasm-bindgen/issues/1742

But seems like it's stalled out.

Reply:
Ah, it would be so nice if this went through. In the meantime, I guess we can keep the boilerplate. And maybe figure out a macro/helper to make it a bit more ergonomic.